### ISSUE TITLE
STAR fails with segmentation fault when parsing GTF

### ISSUE THREAD

**Issue by sidorov-si** [alexdobin/STAR#963]
Dear Alex,

I tested STAR `2.7.5a`, `2.7.3a` and `2.6.1d` on a small dataset with a GTF file in the following way:

```
STAR   \
  --runThreadN 1   \
  --genomeDir 2.7.5a   \
  --readFilesIn ENCFF282NGP_chr6_3400000_3500000_1000reads_1.fq.bz2   ENCFF282NGP_chr6_3400000_3500000_1000reads_2.fq.bz2  \
  --readFilesCommand bunzip2   -c   \
  --outFileNamePrefix Sample1.   \
  --outFilterMultimapNmax 20   \
  --sjdbGTFfile Annotation.gtf
```

where `Annotation.gtf` can be one of the three small example GTFs below:

- gencode.v30.primary_assembly.annotation_chr6_34000000_35000000_**first**_gene.gtf

- gencode.v30.primary_assembly.annotation_chr6_34000000_35000000_**second**_gene.gtf

- gencode.v30.primary_assembly.annotation_chr6_34000000_35000000_**first_and_second**_genes.gtf

(please find them attached: [gtf.zip](https://github.com/alexdobin/STAR/files/4885391/gtf.zip)).

Here "first_gene" and "second_gene" are files with two different genes (one gene per file), and "first_and_second_genes" is a file containing both of these genes.

When I run STAR with these GTFs, I get the following results:

- **Using "first_gene"**: 2.6.1d runs fine, while 2.7.3a and 2.7.5a fail with a segmentation fault at the GTF parsing stage.

- **Using "second_gene"**: all the three versions run fine.

- **Using "first_and_second_genes"**: all the three versions fail with a segmentation fault at the GTF parsing stage.

Attached please find the logs for the 9 runs ([star_run_logs.zip](https://github.com/alexdobin/STAR/files/4885404/star_run_logs.zip)).

Exit status of all failures is `139`.

Could you possibly look into this? 

Is there anything wrong with my GTFs?

Thank you,
     Slava

**Comment by sidorov-si** [comment]
Dear Alex,

I experimented a bit more and have some additional info:

1. As the reference genome I use here is very small (just 100 Kbp), I tried to re-generate it with `--genomeSAindexNbases 7`, as it should be more appropriate according to the manual, and ran STAR 2.6.1d with the "first_gene" GTF on it. Now it failed with a segmentation fault, although worked before on the same reference genome generated with `--genomeSAindexNbases 9` (attached please find the log: [Sample1.Log.out.gz](https://github.com/alexdobin/STAR/files/4902376/Sample1.Log.out.gz)).
 
2. When generating an index with STAR 2.6.1c from the same reference genome using the "first_gene" GTF, it fails with a segmentation fault (please find the logs attached; I tried both `--genomeSAindexNbases 9` and `--genomeSAindexNbases 7`: [genome_generating_with_gtf_logs.zip(https://github.com/alexdobin/STAR/files/4902480/genome_generating_with_gtf_logs.zip)), although it generates the same index just fine if the GTF is not provided. 

3. If I run STAR 2.6.1d on the same test dataset with the "first_gene" GTF, it maps reads successfully (as I mentioned in the previous post), but when I run it a second time using the splice junctintions that it gave me after the first run, it fails with a segmentation fault (attached please find the logs and the `.tab`-file with the junctions: [first_and_second_runs.gz](https://github.com/alexdobin/STAR/files/4902594/first_and_second_runs.gz)).

Overall, I guess, my test dataset could be just too small, but it would be nice, if STAR could finish up gracefully and provide an error message instead of failing with a segmentation fault.

I would really appreciate your help!

All the best,
     Slava

**Comment by alexdobin** [comment]
Hi Slava,

there might be several issues here, let's try to sort them one by one.

Could you please make a test example for a failed run that includes:
Genome.fasta, Annot.gtf, Read1 and Read2. A minimal set of reads that fails is enough.
Also, please include Log.out file for the failed run.

Thanks!
Alex

**Comment by sidorov-si** [comment]
Hi Alex,

Sure, attached please find the FASTA file, the GTF, the paired-end reads and the log of the failed run: [min_failing_dataset.tar.gz](https://github.com/alexdobin/STAR/files/4943925/min_failing_dataset.tar.gz)

I hope, this dataset is reasonably minimal.

The version of STAR is `2.7.5a`. I made the STAR index from the attached FASTA with the following command:

```
STAR   \
  --runMode genomeGenerate   \
  --runThreadN 2   \
  --genomeDir 2.7.5a   \
  --genomeFastaFiles ../raw_genome/GRCh38.primary_assembly.genome_chr6_34000000_35000000.fa     \
  --genomeSAindexNbases 9   \
  --outFileNamePrefix 2.7.5a
```

Thank you,
     Slava

**Comment by alexdobin** [comment]
Hi Slava,

the issue here is that the exon coordinates in the GTF file still refer to the full chromosome, while the fasta starts from the position 34000000. You would need to subtract (34000000-1) from the exon start/end in the GTF. I will fix the code so that STAR ignores exons whose start > chromosome length.

Cheers
Alex

**Comment by sidorov-si** [comment]
Hi Alex,

Oh, thank you so much for the clarification! I'll just change the coordinates then.

Cheers,
     Slava

**Comment by sidorov-si** [comment]
I corrected the coordinates, and it worked indeed! Thanks again!

**Comment by alexdobin** [comment]
Hi Slava,

thanks a lot for reporting this problem. I have fixed it in the 2.7.5b release, now it will skip exons that extend pass the chromosome end, and generate a warning.

Cheers
Alex
